File: May-18/13-05-18/python/02-Unreachable-Nodes.py
N = 10**5 + 5
adj = {}
visit = [0]*N

def dfs(s):
    q = []
    q.append(s)
    visit[s] = 1

    while(len(q) > 0):
        val = q[-1]
        q.pop()
        
        for i in adj[val]:
            if visit[i] == 0:
                visit[i] = 1
                q.append(i)


n, e = map(int, input().split())

for i in range(0,e):
    x, y = map(int, input().split())
    tmp = []
    if x in adj:
        tmp = adj[x]
    tmp.append(y)
    adj[x] = tmp;
    
    tmp = []
    if y in  adj:
        tmp = adj[y]
    tmp.append(x)
    adj[y] = tmp

start = int(input())

dfs(start)

cnt = 0
for i in range(1, n+1):
    if(visit[i] == 0):
        cnt += 1
        
print(cnt)

File: May-18/13-05-18/java/02-Unreachable-Nodes.java

import java.util.*;

class TestClass {
    static int N = (int)1e5;
    static boolean visit[] = new boolean[N];
    static LinkedList<Integer> adj[] = new LinkedList[N];
    
    static void dfs(int s)
    {
        Stack<Integer> st = new Stack<Integer>();
        st.push(s);
        visit[s] = true;
        
        while(st.empty() == false)
        {
            int v = st.pop();
            
            
            Iterator<Integer> i = adj[v].listIterator();
            while (i.hasNext())
            {
                int val = i.next();
                if (visit[val] == false)
                {
                    visit[val] = true;
                    st.push(val);
                }
            }
        }
    }
    public static void main(String args[] ) throws Exception {
        Scanner input = new Scanner(System.in);
        int n, e;
        n = input.nextInt();
        e = input.nextInt();
        
        for(int i = 0; i <= n; i++)
            adj[i] = new LinkedList();
        for(int i = 0; i < e; i++)
        {
            int x, y;
            x = input.nextInt();
            y = input.nextInt();
            adj[x].add(y);
            adj[y].add(x);
        }
        int start = input.nextInt();
        dfs(start);
        
        int cnt = 0;
        for(int i = 1; i <= n; i++)
        {
            if(visit[i] == false)
                cnt += 1;
        }
        System.out.println(cnt);
        
    }
}
